File: golang/src/detectors/xml-external-entity/XmlExternalEntity.go
package main

import (
	"fmt"
	"github.com/lestrrat-go/libxml2/parser"
)

// {fact rule=xml-external-entity@v1.0 defects=1}
func xmlExternalEntityNoncompliant() {
	const xmlData = "<!DOCTYPE d [<!ENTITY e SYSTEM \"file:///etc/passwd\">]><t>&e;</t>"
	// Noncompliant: `XMLParseNoEnt` is enabled.
	xmlParser := parser.New(parser.XMLParseNoEnt)
	parsedDoc, err := xmlParser.ParseString(xmlData)
	if err != nil {
		fmt.Println(err)
		return
	}
	fmt.Println("XML document successfully parsed!")
	fmt.Println(parsedDoc)
}
// {/fact}

// {fact rule=xml-external-entity@v1.0 defects=0}
func xmlExternalEntityCompliant() {
	const xmlData = "<!DOCTYPE d [<!ENTITY e SYSTEM \"file:///etc/passwd\">]><t>&e;</t>"
	// Compliant: `XMLParseNoEnt` is not enabled.
	xmlParser := parser.New()
	parsedDoc, err := xmlParser.ParseString(xmlData)
	if err != nil {
		fmt.Println(err)
		return
	}
	fmt.Println("XML document successfully parsed!")
	fmt.Println(parsedDoc)
}
// {/fact}

File: ruby/src/detectors/xml-external-entity/xml-external-entity.rb
=begin
Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0
=end

# {fact rule=xml-external-entity@v1.0 defects=1}
def xml_external_entity_noncompliant()
  # Noncompliant: Disabling encryption of sensitive data
  config.force_ssl = false
end
# {/fact}

# {fact rule=xml-external-entity@v1.0 defects=0}
def xml_external_entity_compliant()
  # Compliant: Sensitive data is encrypted
  config.force_ssl = true
end
# {/fact}
